grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -155,33 +195,77 @@ class ProductAvailabilityService {
+        // cleanse bin locations, to ensure that we accidentally don't overwrite
+        // (in case INVENTORY or PRODUCT_INVENTORY transaction had entries with other bins)
+        if (binLocations) {
+            binLocationsWithQuantity = binLocationsWithQuantity.findAll { binLocations*.id?.contains(it.id) }
+        }
+        def picked = getQuantityPickedByProductAndLocation(location, product)
+        return transformBinLocations(binLocationsWithQuantity, picked)
+    }
+
+    /**
+     * Get quantity allocated for bin locations and inventory items for a specific location (origin) and product.
+     * Data is pulled from:
+     *  1. Picklist items from pending requisitions (outbound stock movements) with origin being provided location
+     *     that are *NOT* having RECALLED inventory item (inventoryItem.lotStatus) or bin location *WITHOUT*
+     *     HOLD_STOCK in the supported activity (location.supportedActivities).
+     *  2. Picklist items from pending orders (outbound returns) with origin being provided location.
+     *     (IMPORTANT: Outbound returns can have picked items with RECALLED lots and bins with HOLD_STOCK activity)
+     * */
     def getQuantityPickedByProductAndLocation(Location location, Product product) {
-        def results = Picklist.executeQuery("""
-            SELECT
-                pli.binLocation,
-                pli.inventoryItem,
-                sum(pli.quantity)*((count(distinct pli.requisitionItem) + count(distinct pli.orderItem) )/ count(*))
-            FROM PicklistItem pli
-            INNER JOIN pli.picklist pl
-            LEFT JOIN pli.requisitionItem ri
-            LEFT JOIN pl.requisition r
-            LEFT JOIN pli.orderItem oi
-            LEFT JOIN pl.order o
-            LEFT JOIN pli.inventoryItem ii
-            LEFT JOIN pli.binLocation l
-            LEFT JOIN l.supportedActivities s
-            WHERE ((r.origin = :location AND r.status IN (:pendingRequisitionStatus)) OR (o.origin = :location AND o.status IN (:pendingOrderStatus)))
-              AND (oi IS NOT NULL OR (ri IS NOT NULL AND (ii.lotStatus IS NULL OR ii.lotStatus != 'RECALLED') AND NOT ('HOLD_STOCK' IN ELEMENTS(l.supportedActivities))))
-              AND (:product = '' OR ri.product.id = :product OR oi.product.id = :product)
-            GROUP BY pli.binLocation, pli.inventoryItem
-        """, [
-                location                    : location,
-                pendingRequisitionStatus    : RequisitionStatus.listPending(),
-                pendingOrderStatus          : OrderStatus.listPending(),
-                product                     : product?.id ?: ''
-        ], [readOnly: true])
-
-        return results.collect { [binLocation: it[0]?.id, inventoryItem: it[1]?.id, quantityAllocated: it[2]] }
+        def query = """
+            SELECT bin_location_id, inventory_item_id, SUM(quantity) FROM (
```

Comment:
As I mentioned on the call the other day, we should include product and location in the select even though they are passed in as query parameters. There's a small chance that might lead to slower performance if those columns aren't in a covering index but I don't suspect that's the case. Also, the inner (raw query) should not include the sum / group by if we can help it. It's usually better to leave the group by until the end i.e. outer query. I could be wrong about this, but it feels wrong to apply aggregation inside a nested query if you're going to do the same on the outside. 

Also I see we're getting rid of the weird count + count / count(*). I think that's safe since we're simplifying the query by breaking into two separate queries and unioning the results. But we should try to understand and/or compare results from the previous query so we know we're getting the same results.